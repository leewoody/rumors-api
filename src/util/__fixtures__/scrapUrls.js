export default {
  '/urls/doc/someUrl': {
    url: 'http://example.com/article/1111-aaaaa-bbb-ccc',
    canonical: 'http://example.com/article/1111',
    title: 'Example title',
    summary: 'Extracted summary',
    fetchedAt: new Date('2017-01-01'),
  },
  '/urls/doc/someUrl-2nd-fetch': {
    url: 'http://example.com/article/1111-aaaaa-bbb-ccc',
    canonical: 'http://example.com/article/1111',
    title: 'Changed title',
    summary: 'Changed summary',
    fetchedAt: new Date('2017-02-01'),
  },
};
