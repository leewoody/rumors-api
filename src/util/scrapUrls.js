/* eslint-env browser */

import puppeteer from 'puppeteer';
import path from 'path';
import fs from 'fs';

const readabilityJsStr = fs.readFileSync(
  path.join(__dirname, '../../node_modules/readability/Readability.js'),
  { encoding: 'utf-8' }
);

/* eslint-disable no-undef */
/**
 * Executed in puppeteer browser
 */
function executor() {
  return new Readability({}, document).parse();
}
/* eslint-enable no-undef */

/**
 * Return type for scrapUrls
 * @typedef {Object} ScrapResult
 * @property {string} canonical Canonical URL
 * @property {string} title
 * @property {string} summary
 * @property {string} html
 * @property {string} topImageUrl
 */

/**
 * Fetches the given url
 * @param {browser} browser - Puppeteer browser instance
 * @param {string} url - The URL to scrap
 * @return {Promise<ScrapResult>}
 */
async function scrap(browser, url) {
  const page = await browser.newPage();

  page.setDefaultNavigationTimeout(5000);
  try {
    await page.goto(url);
  } catch (e) {
    // Something like timeout. Do nothing.
  }

  const html = await page.content();

  const canonical = await page.evaluate(() => {
    const canonicalLink = document.querySelector('link[rel=canonical]');
    if (canonicalLink) return canonicalLink.href;

    const ogUrlMeta = document.querySelector('meta[property="og:url"]');
    if (ogUrlMeta) return ogUrlMeta.content;

    return window.location.href;
  });

  const resultArticle = await page.evaluate(`
    (function(){
      ${readabilityJsStr}
      ${executor}
      return executor();
    }())
  `);

  // If the whole page is empty due to any error, just return null
  if (!resultArticle) return null;

  const topImageUrl = await page.evaluate(contentHTML => {
    const ogImageMeta = document.querySelector(
      'meta[property="og:image"], meta[property="og:image:url"]'
    );
    if (ogImageMeta) return ogImageMeta.content;

    const containerDiv = document.createElement('div');
    containerDiv.innerHTML = contentHTML;
    const contentImgs = Array.from(containerDiv.querySelectorAll('img'));
    if (contentImgs.length === 0) return '';

    const largestImg = contentImgs.slice(1).reduce((largestImage, img) => {
      return largestImage.width * largestImage.height >= img.width * img.height
        ? largestImage
        : img;
    }, contentImgs[0]);

    // src may be relative URL, resolve with current location.
    return new URL(largestImg.src, location.href).href;
  }, resultArticle.content);

  return {
    canonical,
    title: resultArticle.title,
    summary: resultArticle.textContent.trim(),
    topImageUrl,
    html,
  };
}

/**
 * Fetches the specified url data from cache or through scrapping
 *
 * @param {string[]} urls
 * @param {object} options
 * @param {boolean} options.cacheLoader - The dataloader that loads result from `urls`, given a URL.
 *                                        If not given, scrapUrls don't read `urls` cache.
 * @param {boolean} options.noFetch - If true, return null when cache does not hit. Default: false.
 * @return {Promise<ScrapResult[]>} - If cannot scrap, resolves to null
 */
async function scrapUrls(urls, { cacheLoader, noFetch = false } = {}) {
  let browserPromise;

  const result = await Promise.all(
    urls.map(async url => {
      if (cacheLoader) {
        const result = await cacheLoader.load(url);
        if (result) return result;
      }

      if (noFetch) return null;

      // Lazily instantiate puppeteer
      if (!browserPromise) browserPromise = puppeteer.launch();

      return scrap(await browserPromise, url);
    })
  );

  // Cleanup browser if opened
  if (browserPromise) {
    (await browserPromise).close();
  }

  return result;
}

export default scrapUrls;
