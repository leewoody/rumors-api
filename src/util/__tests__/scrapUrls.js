import Koa from 'koa';
import path from 'path';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import fixtures from '../__fixtures__/scrapUrls';
import koaStatic from 'koa-static';
import scrapUrls from '../scrapUrls';
import DataLoaders from 'graphql/dataLoaders';
const PORT = 57319;

describe('scrapping', () => {
  let server;
  beforeAll(
    async () =>
      new Promise(resolve => {
        const app = new Koa();
        app.use(koaStatic(path.join(__dirname, '../__fixtures__/server/')));
        server = app.listen(PORT, resolve);
      })
  );

  afterAll(() => server.close());

  it('scraps from Internet and handles error', async () => {
    const [
      { html, ...foundResult },
      notFoundResult,
      invalidUrlResult,
    ] = await scrapUrls([
      `http://localhost:${PORT}/index.html`,
      `http://localhost:${PORT}/not-found`,
      'this is invalid URL',
    ]);

    expect(html.slice(0, 100)).toMatchSnapshot('foundResult html'); // Too long, just sample first 100 chars
    expect(foundResult).toMatchSnapshot('foundResult');
    expect(notFoundResult).toMatchSnapshot('notFoundResult');
    expect(invalidUrlResult).toBeNull();
  });
});

describe('caching', () => {
  beforeAll(() => loadFixtures(fixtures));
  afterAll(() => unloadFixtures(fixtures));

  it('fetches latest fetch from urls', async () => {
    const loaders = new DataLoaders();
    expect(
      await scrapUrls(
        [
          'http://example.com/article/1111-aaaaa-bbb-ccc', // full URL
          'http://example.com/article/1111', // canonical URL
        ],
        {
          cacheLoader: loaders.urlLoader,
          noFetch: true,
        }
      )
    ).toMatchSnapshot();
  });
});
