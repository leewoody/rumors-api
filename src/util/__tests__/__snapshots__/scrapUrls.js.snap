// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`caching fetches latest fetch from urls 1`] = `
Array [
  Object {
    "_cursor": undefined,
    "_score": 0.18232156,
    "canonical": "http://example.com/article/1111",
    "fetchedAt": "2017-01-01T00:00:00.000Z",
    "id": "someUrl",
    "summary": "Extracted summary",
    "title": "Example title",
    "url": "http://example.com/article/1111-aaaaa-bbb-ccc",
  },
  Object {
    "_cursor": undefined,
    "_score": 0.18232156,
    "canonical": "http://example.com/article/1111",
    "fetchedAt": "2017-02-01T00:00:00.000Z",
    "id": "someUrl-2nd-fetch",
    "summary": "Changed summary",
    "title": "Changed title",
    "url": "http://example.com/article/1111-aaaaa-bbb-ccc",
  },
]
`;

exports[`scrapping scraps from Internet and handles error: foundResult 1`] = `
Object {
  "canonical": "http://localhost:57319/index.html",
  "summary": "2018-02-09 16:46中央社 台北9日電大同公司今天聲明，近期大同公司或大同3C並無贈送LINE貼圖或對外募款活動，相關詐騙訊息皆非由大同或其子公司官方發出，建議消費者查證並提高警覺。大同發布聲明指出，近期LINE出現盜用大同公司官方名義與大同寶寶圖像進行假訊息散布，並發送花蓮地震救災捐款訊息，且以優惠活動、贈送貼圖等手段，騙取加入好友、分享貼圖，藉此竊取個資。針對上述冒用詐騙情事，大同表示，大同3C日前已於其官網張貼警示昭告，大同公司也嚴正聲明，近期大同公司或大同3C並無贈送LINE貼圖或對外募款的活動。大同強調，上述詐騙訊息皆非由大同或其子公司官方發出，消費者若接獲活動訊息，請先至大同公司或大同3C官方平台查詢，包含官方網站、臉書粉絲頁、LINE官方帳號等，或洽大同官方免付費服務專線0800-052-666進一步求證。大同籲請社會大眾對社群媒體資訊提高警覺，並直接將詐騙訊息刪除，切勿點擊或轉發，以保障個人資料安全並避免損失。對於冒用大同公司、大同3C官方名義以及大同寶寶圖像者，大同公司表示必將追究，絕不寬貸。大同公司表示，近期LINE出現盜用大同官方帳號進行假訊息散佈，請多加注意，以免受騙。圖／擷自大同3C官網。分享",
  "title": "Test document",
  "topImageUrl": "https://pgw.udn.com.tw/gw/photo.php?u=https://uc.udn.com.tw/photo/2018/02/09/98/4490266.jpg&x=0&y=0&sw=0&sh=0&sl=W&fw=1050&exp=3600",
}
`;

exports[`scrapping scraps from Internet and handles error: foundResult html 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"widt"
`;

exports[`scrapping scraps from Internet and handles error: notFoundResult 1`] = `
Object {
  "canonical": "http://localhost:57319/not-found",
  "html": "<html><head></head><body><pre style=\\"word-wrap: break-word; white-space: pre-wrap;\\">Not Found</pre></body></html>",
  "summary": "Not Found",
  "title": "",
  "topImageUrl": "",
}
`;
